pkg/boot/universalpayload/utilities.go

```diff
@@ -170,3 +225,144 @@ func alignHOBLength(expectLen uint64, bufLen int, buf *bytes.Buffer) error {
 	}
 	return nil
 }
+
+// Constrcut trampoline code before jump to entry point of FIT image.
+// Due to lack of support to set value of General Purpose Registers in kexec,
+// bootloader parameter needs to be prepared in trampoline code.
+// Also stack is prepared in trampoline code snippet to ensure no data leak.
+//
+// Trampoline code snippet is prepared as following:
+//
+//	trampoline[0 - 6]   : mov rax, qword ptr [rip+0x19]
+//	trampoline[7 - 9]   : mov rsp, rax
+//	trampoline[10 - 16] : mov rax, qword ptr [rip+0x17]
+//	trampoline[17 - 19] : mov rcx, rax
+//	trampoline[20 - 26] : mov rax, qword ptr [rip+0x15]
+//	trampoline[27 - 28] : jmp rax
+//	trampoline[29 - 31] : padding for alignment
+//	trampoline[32 - 39] : Top of stack address
+//	trampoline[40 - 47] : Base address of bootloader parameter
+//	trampoline[48 - 55] : Entry point of FIT image
+func constructTrampoline(buf []uint8, hobAddr uint64, entry uint64) []uint8 {
+	loadStackAddress := []uint8{0x48, 0x8b, 0x05, 0x19, 0x00, 0x00, 0x00}
+	setStackAddress := []uint8{0x48, 0x89, 0xc4}
+	loadBootparameter := []uint8{0x48, 0x8b, 0x05, 0x17, 0x00, 0x00, 0x00}
+	setBootparameter := []uint8{0x48, 0x89, 0xc1}
+	loadKernelAddress := []uint8{0x48, 0x8b, 0x05, 0x15, 0x00, 0x00, 0x00}
+	jumpToKernelAddress := []uint8{0xff, 0xe0}
+	padForAlignment := []uint8{0x00, 0x00, 0x00}
+
+	buf = append(buf, loadStackAddress...)
+	buf = append(buf, setStackAddress...)
+	buf = append(buf, loadBootparameter...)
+	buf = append(buf, setBootparameter...)
+	buf = append(buf, loadKernelAddress...)
+	buf = append(buf, jumpToKernelAddress...)
+	buf = append(buf, padForAlignment...)
+
+	stackTop := hobAddr + tmpStackTop
+	appendUint64 := func(slice []uint8, value uint64) []uint8 {
+		tmpBytes := make([]uint8, 8)
+		binary.LittleEndian.PutUint64(tmpBytes, value)
+		return append(slice, tmpBytes...)
+	}
+
+	buf = appendUint64(buf, stackTop)
+	buf = appendUint64(buf, hobAddr)
+	buf = appendUint64(buf, entry)
+
+	return buf
+}
+
+// Walk through .reloc section, update expected address to actual address
+// which is calculated with recloation offset. Currently, only type of
+// IMAGE_REL_BASED_DIR64(10) found in .reloc setcion, update this type
+// of address only.
+func relocatePE(relocData []byte, delta uint64, data []byte) error {
+	r := bytes.NewReader(relocData)
+
+	for {
+		// Read relocation block header
+		var pageRVA uint32
+		var blockSize uint32
+
+		err := binary.Read(r, binary.LittleEndian, &pageRVA)
+		if err == io.EOF {
+			break // End of relocations
+		}
+		if err != nil {
+			return fmt.Errorf("%w: err: %w", ErrPeFailToGetPageRVA, err)
+		}
+		err = binary.Read(r, binary.LittleEndian, &blockSize)
+		if err != nil {
+			return fmt.Errorf("%w: err: %w", ErrPeFailToGetBlockSize, err)
+		}
```

Comment:
nit: simplify error checking
```suggestion
		if err = binary.Read(r, binary.LittleEndian, &blockSize); err != nil {
			return fmt.Errorf("%w: err: %w", ErrPeFailToGetBlockSize, err)
		}
```